### DISCUSSION TITLE
Support bind-mounts for npm ci

### DISCUSSION THREAD

**Discussion by retrohacker** [npm/feedback#73] [CLI] [answered]
Today `npm install` can modify the `package-lock.json` file. To honor a lock file, i.e. for CI runs, we invoke `npm ci`. `npm ci` attempts to `rm -rf` the `node_modules` directory. But the `node_modules` "directory" isn't always rimraf-able.

For example, during a CI run, you may want to capture the side effects of an `npm ci` and re-use that snapshot if the hash of the package-lock.json matches a future CI run. To do this, you may bind-mount some permanent form of storage to the `node_modules` directory so it outlives the lifecycle of the single CI run. In order to support this, the empty `node_modules` directory _must_ exist before invoking `npm ci` and `npm ci` must not try to delete it.

This flow is very common when running `npm` inside of docker containers.

More context and conversation here: https://github.com/npm/cli/issues/564

**Comment by ljharb** [comment]
I've run into this myself - in my case, I have my Dropbox set up to selectively ignore syncing `node_modules` dirs, but when the dir gets deleted, Dropbox loses this setting (i believe it's set with xattrs on the dir on my Mac); it would be *really* nice if instead of `rimraf node_modules` it did `rimraf node_modules/{.*,*}`.

**Comment by MylesBorins** [comment]
A couple thoughts here

I believe you can run `npm install --no-save` which should install, no blow away node_modules, and respect package-lock.json

[A comment from Claudia](https://github.com/npm/cli/issues/564#issuecomment-564050930) in the issue you shared seems to imply that the CLI may have at some point stopped deleting the entire directory and switched to deleting the contents... this is potentially something that we regressed on an we are looking into it right now.

**Comment by nlf** [comment, marked as answer]
typically we would suggest that people persist the npm cache between installs rather than `node_modules`, which makes `npm ci` really fast as it avoids the vast majority of network requests

if your use case is to artifact and persist the entire app as it is, is there a need to even run `npm install` or `npm ci` again at all?

that said, as @MylesBorins mentioned it is definitely reasonable for us to delete the contents of `node_modules` but not the directory, we can get that patched up for sure

**Reply by nlf** [reply]
here's a link to the PR that fixes the regression and deletes only the contents of `node_modules` and not the directory itself https://github.com/npm/cli/pull/2083

**Reply by retrohacker** [reply]
:raised_hands: you rock, thank you!

In case your curious, I'll still answer the questions :heart:

To slightly oversimplify: we are able to skip invoking `npm` entirely in our build graph assuming we have an input (package-lock.json) that maps to an output (node_modules) already. It's independent of the artifact and build. That mapping is being persisted in a cache that all builds have access to so, in the future, if we come across that same node in another build graph we can reuse the same outputs. It also allows us to handle that branch of the build graph in parallel with other branches (an `npm ci` can be happening at the same time as an `apt install` for the same artifact).
